### PR TITLE
[#175761632] Add CGN Activation API

### PR DESCRIPTION
**Pull request by AleDore**
#### List of Changes
- Add `StartCgnActivation` API
- Add `StoreCgnExpirationActivity`
- Add OpenAPI endpoint spec

#### Motivation and Context
A CGN can be activated only for citizen between 18 and 35 years old, so IO App should have an API to be called in order to perform these steps:
- Perform "Eligibility Check" for citizen (by checking his/her age through `fiscalCode`)
- Check if there are other update processes that could impact the same CGN
- Store a `UserCgn` record in status PENDING to recognize an activation process is running
- Store the expiration of the CGN in a proper table, in order to be processed by CgnExpiration trigger, see [https://github.com/pagopa/io-functions-cgn/pull/11](https://github.com/pagopa/io-functions-cgn/pull/11)
- Update CGN status to `ACTIVATED`
- Send proper message to citizen with all the infos about CGN's status change

There's other steps that should be considered in this process and will be implemented into an other PR (EYCA API integration).
An overall overview of the activation process can be resumed as following:

CGN - Activation Flow

Please refer to related [story](https://www.pivotaltracker.com/story/show/175761632) for further informations.

#### How Has This Been Tested?
It has been tested by performing unit tests.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Review by balanza — CHANGES_REQUESTED** [review]
Besides the little comments I left, I have doubts about the way we infer CGN status.

Do we need to check the orchestrator? Can't we have an explicit status, updated before and after an operation is performed on a CGN?

**Comment by AleDore** [comment]
> Besides the little comments I left, I have doubts about the way we infer CGN status.
> 
> Do we need to check the orchestrator? Can't we have an explicit status, updated before and after an operation is performed on a CGN?

Maybe it is more clear now after [this](https://github.com/pagopa/io-functions-cgn/pull/13/commits/353840e957170cb13621c18449b4da816f7bb0ca)

File: StartCgnActivation/handler.ts [hunk 1]
```diff
@@ -117,63 +121,61 @@ export function StartCgnActivationHandler(
       ActivatedStatusEnum.ACTIVATED
     ) as NonEmptyString;
 
-    const isEligibleResponseOrError = await checkCgnEligibleDataTask(
+    const isExpirationCgnOrError = await getCgnExpirationDataTask(
```
Comment (balanza) [inline StartCgnActivation/handler.ts:124]: Nitpicking: we should use `is-` and `has-` prefixes only for booleans

**Comment by balanza** [comment]
> > Besides the little comments I left, I have doubts about the way we infer CGN status.
> > Do we need to check the orchestrator? Can't we have an explicit status, updated before and after an operation is performed on a CGN?
> 
> @balanza Maybe it is more clear now after [this](https://github.com/pagopa/io-functions-cgn/pull/13/commits/353840e957170cb13621c18449b4da816f7bb0ca)

I don't understand. Aren't we still inferring the status of the process by the status of the orchestrator?

**Comment by AleDore** [comment]
> > > Besides the little comments I left, I have doubts about the way we infer CGN status.
> > > Do we need to check the orchestrator? Can't we have an explicit status, updated before and after an operation is performed on a CGN?
> > 
> > 
> > @balanza Maybe it is more clear now after [this](https://github.com/pagopa/io-functions-cgn/pull/13/commits/353840e957170cb13621c18449b4da816f7bb0ca)
> 
> I don't understand. Aren't we still inferring the status of the process by the status of the orchestrator?

Introduction:
A CGN activation process can be resumed as a state machine with only 2 states: `PENDING` and `ACTIVATED`. When a new CGN activation starts, it creates a UserCgn with `PENDING` status and while the orchestrator is running there's an activity that updates the CGN status to `ACTIVATED`. Please note that if an orchestrator was terminated or canceled or it raise an error, then CGN status couldn't be updated to ACTIVATED. Another info to take care about is: every orchestrator has a persistent tier that allows durable functions to be restored, or even to track its execution history. These informations, after a time-boxed inactivity (default 30 days if I' m not wrong), could be deleted, so after this time you cannot retrieve any execution information about its status.

After the refactor, the logic to infer the status is:
- Check if there is a UserCgn model on cosmos
- If present and CGN status is in a final status, then it makes no sense to start an activation so we will return a proper error response (namely 409)
- If not present we are sure that no other activation has been triggered before so we can move on
- if a CGN is in PENDING Status, we have to check if the related activation orchestrator is in a consistent status, in order to discriminate if we can move on with the activation ( even a new activation after a failed one ), or if we have to return an error.

That's all folks!
What do you think @balanza ?

**Comment by balanza** [comment]
Ok, so I got it wrong: **CGN status is not inferred by the status of the orchestrator**, which is actually checked only to prevent duplicated execution. Am I right?

**Comment by AleDore** [comment]
> Ok, so I got it wrong: **CGN status is not inferred by the status of the orchestrator**, which is actually checked only to prevent duplicated execution. Am I right?

Yep, or even this is the status of the **Activation process** related to a CGN.

**Review by balanza — APPROVED** [review]
LGTM, thanks for the deep explanation

**Pull request merged**